### PR TITLE
feat: use vision to add mask

### PR DESCRIPTION
**Pull request by andrewlew1s**
Completed Work:

- Set up Google Cloud Project & credentials (in discord)
- Got facial data from Vision API 
- Deployed cloud function to receive said data
- Integrated with UI
- Did some basic maths

Screen Shot 2020-05-12 at 1 49 24 AM
Screen Shot 2020-05-12 at 1 49 19 AM
Screen Shot 2020-05-12 at 1 49 13 AM

ToDo:

- More maths

### CODE REVIEW

File: cloud-functions/functions/index.js [hunk 1]
```diff
@@ -1,4 +1,15 @@
...
-exports.getFacialRecognitionData = functions.https.onCall(url => `I can't find a face in ${url}`);
-// TODO make a call to google vision API with image url
+const vision = require('@google-cloud/vision');
+
+const client = new vision.ImageAnnotatorClient();
+
+exports.getFacialRecognitionData = functions.https.onCall(async url => {
+	try {
+		const [result] = await client.faceDetection(url);
+		console.log(JSON.stringify(result, null, '\t'));
```
Comment (michaelfitzhavey) [inline cloud-functions/functions/index.js:10]: Let's remove this line before deploying

File: website/src/components/ImageMasker.vue [hunk 2]
```diff
@@ -20,7 +20,7 @@
 						v-model="maskSizing.leftOffset"
 						type="range"
 						min="0"
-						max="1"
+						max="2"
```
Comment (michaelfitzhavey) [inline website/src/components/ImageMasker.vue:23]: Why have we modified `max`?

File: website/src/components/ImageMasker.vue [hunk 3]
```diff
@@ -98,19 +98,19 @@ export default {
 			required: true
 		},
 		data: {
-			type: String, // TODO: switch type to object and use facial recognition data to add mask
-			requied: true
+			type: Object, // TODO: switch type to object and use facial recognition data to add mask
```
Comment (michaelfitzhavey) [inline website/src/components/ImageMasker.vue:101]: We should remove the (now incorrect) comment

File: website/src/components/ImageMasker.vue [hunk 4]
```diff
@@ -98,19 +98,19 @@ export default {
...
 	data() {
 		return {
 			context: null,
 			originalDrawn: false,
 			maskSizing: {
-				leftOffset: 0.3,
-				topOffset: 0.4,
-				width: 0.5,
-				height: 0.2
+				leftOffset: 1,
```
Comment (michaelfitzhavey) [inline website/src/components/ImageMasker.vue:110]: These default values are useful for if no face is found, why are we deleting them?

File: website/src/components/ImageMasker.vue [hunk 5]
```diff
@@ -125,12 +125,43 @@ export default {
 			);
 		},
 		drawMask(mask) {
```
Comment (michaelfitzhavey) [inline website/src/components/ImageMasker.vue:127]: I think the drawmask functionality should remain as it was, perhaps we need a computed value `defaultMaskPosition` that uses `imageData`.

It's important we don't confuse the functionality of working out the bounding box and rotation of the ideal mask with the functionality of overlaying one PNG on another (what `drawMask` currently does).

File: website/src/components/ImageMasker.vue [hunk 6]
```diff
@@ -125,12 +125,43 @@ export default {
...
+					mouthCentreY = landmark.position.y;
+				}
+			});
+			width = rightEarX - leftEarX;
+			height = chinLowest - eyeMiddle;
+			toOffsetLeft = this.imageSize - rightEarX;
+			toOffsetTop = this.imageSize - mouthCentreY;
 			this.context.drawImage(
 				this.$refs[mask],
-				this.imageSize * this.maskSizing.leftOffset,
```
Comment (michaelfitzhavey) [inline website/src/components/ImageMasker.vue:130]: `this.imageSize` matters, don't delete all references to it!

**Review by fitzhavey — CHANGES_REQUESTED** [review]

**Comment by michaelfitzhavey** [comment]
You should run `git pull origin master` as https://github.com/holistic-web/maskify/pull/8 has been merged.

**Comment by michaelfitzhavey** [comment]
This PR may be out of date once this is merged: https://github.com/holistic-web/maskify/pull/11

**Pull request closed without merging**